go-controller/pkg/cni/cni_test.go

```diff
@@ -131,27 +136,168 @@ var _ = Describe("Network Segmentation", func() {
+							"k8s.ovn.org/pod-networks": `{"default":{"ip_address":"100.10.10.3/24","mac_address":"0a:58:fd:98:00:01", "role":"primary"}}`,
+						},
+					},
+				}
+			})
+			It("should not fail at cmdAdd", func() {
+				podNamespaceLister.On("Get", pr.PodName).Return(pod, nil)
+				Expect(pr.cmdAddWithGetCNIResultFunc(kubeAuth, clientSet, getCNIResultStub)).NotTo(BeNil())
+				Expect(obtainedPodIterfaceInfos).ToNot(BeEmpty())
+			})
+			It("should not fail at cmdDel", func() {
+				podNamespaceLister.On("Get", pr.PodName).Return(pod, nil)
+				Expect(pr.cmdDel(clientSet)).NotTo(BeNil())
+				Expect(prInterfaceOpsStub.unconfiguredInterfaces).To(HaveLen(2))
+			})
 		})
 
+		Context("pod with a user defined primary network", func() {
+			const (
+				dummyMACHostSide = "07:06:05:04:03:02"
+				nadName          = "tenantred"
+				namespace        = "foo-ns"
+			)
+
+			dummyGetCNIResult := func(request *PodRequest, getter PodInfoGetter, podInterfaceInfo *PodInterfaceInfo) (*current.Result, error) {
+				var gatewayIP net.IP
+				if len(podInterfaceInfo.Gateways) > 0 {
+					gatewayIP = podInterfaceInfo.Gateways[0]
+				}
+				var ips []*current.IPConfig
+				ifaceIdx := 1 // host side of the veth is 0; pod side of the veth is 1
+				for _, ip := range podInterfaceInfo.IPs {
+					ips = append(ips, &current.IPConfig{Address: *ip, Gateway: gatewayIP, Interface: &ifaceIdx})
+				}
+				ifaceName := "eth0"
+				if request.netName != "default" {
+					ifaceName = "ovn-udn1"
+				}
+
+				interfaces := []*current.Interface{
+					{
+						Name: "host_" + ifaceName,
+						Mac:  dummyMACHostSide,
+					},
+					{
+						Name:    ifaceName,
+						Mac:     podInterfaceInfo.MAC.String(),
+						Sandbox: "bobloblaw",
+					},
+				}
+				return &current.Result{
+					CNIVersion: "0.3.1",
+					Interfaces: interfaces,
+					IPs:        ips,
+				}, nil
+			}
+
+			BeforeEach(func() {
+				pod = &v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      pr.PodName,
+						Namespace: pr.PodNamespace,
+						Annotations: map[string]string{
+							"k8s.ovn.org/pod-networks": `{"default":{"ip_address":"100.10.10.3/24","mac_address":"0a:58:fd:98:00:01", "role":"infrastructure-locked"}, "meganet":{"ip_address":"10.10.10.30/24","mac_address":"02:03:04:05:06:07", "role":"primary"}}`,
+						},
+					},
+				}
+				nadNamespaceLister := &v1nadmocks.NetworkAttachmentDefinitionNamespaceLister{}
+				nadNamespaceLister.On("List", labels.Everything()).Return([]*nadv1.NetworkAttachmentDefinition{{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      nadName,
+						Namespace: namespace,
+					},
+					Spec: nadv1.NetworkAttachmentDefinitionSpec{
+						Config: dummyPrimaryUDNConfig(namespace, nadName),
+					},
+				}}, nil)
+				nadLister.On("NetworkAttachmentDefinitions", "foo-ns").Return(nadNamespaceLister)
+
+				getCNIResultStub = dummyGetCNIResult
+			})
+
+			It("should return the information of both the default net and the primary UDN in the result", func() {
```

Comment:
could we have a dual stack test?

Reply:
If you prefer, I'll just turn this single test into dual stack. How about that ?

Reply:
I am fine with that.

Reply:
Done.